File: src/main/java/raylras/zen/langserver/provider/data/Keywords.java
package raylras.zen.langserver.provider.data;

public class Keywords {

    public static final String[] ALL = makeAllKeywords();
    public static final String[] CLASS_BODY = makeClassBodyKeywords();
    public static final String[] LOCAL_STATEMENT = makeLocalStatementKeywords();
    public static final String[] TOPLEVEL_STATEMENT = makeToplevelStatementKeywords();


    public static final String IMPORT = "import";

    public static final String VAR = "var";
    public static final String VAL = "val";
    public static final String GLOBAL = "global";
    public static final String STATIC = "static";
    public static final String FUNCTION = "function";
    public static final String ZEN_CLASS = "zenClass";
    public static final String ZEN_CONSTRUCTOR = "zenConstructor";

    public static final String IF = "if";
    public static final String ELSE = "else";
    public static final String FOR = "for";
    public static final String WHILE = "while";
    public static final String BREAK = "break";
    public static final String CONTINUE = "continue";
    public static final String RETURN = "return";

    public static final String AS = "as";
    public static final String IN = "in";
    public static final String HAS = "has";
    public static final String INSTANCEOF = "instanceof";
    public static final String THIS = "this";
    public static final String TRUE = "true";
    public static final String FALSE = "false";
    public static final String NULL = "null";

    public static final String ANY = "any";
    public static final String BYTE = "byte";
    public static final String SHORT = "short";
    public static final String INT = "int";
    public static final String LONG = "long";
    public static final String FLOAT = "float";
    public static final String DOUBLE = "double";
    public static final String BOOL = "bool";
    public static final String VOID = "void";
    public static final String STRING = "string";

    public static final String EXPAND = "$expand";

    private static String[] makeAllKeywords() {
        return new String[]{
                IMPORT,
                VAR, VAL, GLOBAL, STATIC, FUNCTION, ZEN_CLASS, ZEN_CONSTRUCTOR,
                IF, ELSE, FOR, WHILE, BREAK, CONTINUE, RETURN,
                AS, IN, HAS, INSTANCEOF, THIS, TRUE, FALSE, NULL,
                ANY, BYTE, SHORT, INT, LONG, FLOAT, DOUBLE, BOOL, VOID, STRING,
                EXPAND,
        };
    }

    private static String[] makeClassBodyKeywords() {
        return new String[]{
                VAR, VAL, STATIC, FUNCTION, ZEN_CONSTRUCTOR
        };
    }

    private static String[] makeToplevelStatementKeywords() {
        return new String[]{
                IMPORT,
                VAR, VAL, STATIC, GLOBAL, FUNCTION, ZEN_CLASS,
                IF, ELSE, FOR, WHILE,
                EXPAND
        };
    }

    private static String[] makeLocalStatementKeywords() {
        return new String[]{
                VAR, VAL,
                IF, ELSE, FOR, WHILE, BREAK, CONTINUE, RETURN,
                THIS
        };
    }

}

File: zenscript-language-server/src/main/kotlin/raylras/zen/lsp/provider/data/Keywords.kt
package raylras.zen.lsp.provider.data

object Keywords {
    val ALL: Array<String> = makeAllKeywords()
    val CLASS_BODY: Array<String> = makeClassBodyKeywords()
    val STATEMENT: Array<String> = makeStatementKeywords()
    val LOCAL_STATEMENT: Array<String> = makeLocalStatementKeywords()
    val TOPLEVEL_STATEMENT: Array<String> = makeToplevelStatementKeywords()


    const val IMPORT: String = "import"

    const val VAR: String = "var"
    const val VAL: String = "val"
    const val GLOBAL: String = "global"
    const val STATIC: String = "static"
    const val FUNCTION: String = "function"
    const val ZEN_CLASS: String = "zenClass"
    const val ZEN_CONSTRUCTOR: String = "zenConstructor"

    const val IF: String = "if"
    const val ELSE: String = "else"
    const val FOR: String = "for"
    const val WHILE: String = "while"
    const val BREAK: String = "break"
    const val CONTINUE: String = "continue"
    const val RETURN: String = "return"

    const val AS: String = "as"
    const val IN: String = "in"
    const val HAS: String = "has"
    const val INSTANCEOF: String = "instanceof"
    const val THIS: String = "this"
    const val TRUE: String = "true"
    const val FALSE: String = "false"
    const val NULL: String = "null"

    const val ANY: String = "any"
    const val BYTE: String = "byte"
    const val SHORT: String = "short"
    const val INT: String = "int"
    const val LONG: String = "long"
    const val FLOAT: String = "float"
    const val DOUBLE: String = "double"
    const val BOOL: String = "bool"
    const val VOID: String = "void"
    const val STRING: String = "string"

    const val EXPAND: String = "\$expand"

    private fun makeAllKeywords(): Array<String> {
        return arrayOf(
            IMPORT,
            VAR, VAL, GLOBAL, STATIC, FUNCTION, ZEN_CLASS, ZEN_CONSTRUCTOR,
            IF, ELSE, FOR, WHILE, BREAK, CONTINUE, RETURN,
            AS, IN, HAS, INSTANCEOF, THIS, TRUE, FALSE, NULL,
            ANY, BYTE, SHORT, INT, LONG, FLOAT, DOUBLE, BOOL, VOID, STRING,
            EXPAND,
        )
    }

    private fun makeClassBodyKeywords(): Array<String> {
        return arrayOf(
            VAR, VAL, STATIC, FUNCTION, ZEN_CONSTRUCTOR
        )
    }

    private fun makeStatementKeywords(): Array<String> {
        return arrayOf(
            IMPORT,
            VAR, VAL, STATIC, GLOBAL, FUNCTION, ZEN_CLASS, ZEN_CONSTRUCTOR,
            IF, ELSE, FOR, WHILE, BREAK, CONTINUE, RETURN,
            EXPAND
        )
    }

    private fun makeToplevelStatementKeywords(): Array<String> {
        return arrayOf(
            IMPORT,
            VAR, VAL, STATIC, GLOBAL, FUNCTION, ZEN_CLASS,
            IF, ELSE, FOR, WHILE,
            EXPAND
        )
    }

    private fun makeLocalStatementKeywords(): Array<String> {
        return arrayOf(
            VAR, VAL,
            IF, ELSE, FOR, WHILE, BREAK, CONTINUE, RETURN,
            THIS
        )
    }
}
